src/workerd/api/worker-rpc.h

```diff
@@ -81,6 +81,115 @@ class JsRpcTarget: public jsg::Object {
   JSG_ONEWAY_SERIALIZABLE(rpc::SerializationTag::JS_RPC_STUB);
 };
 
+// Common superclass of JsRpcStub and Fetcher, the two types that may serve as the basis for
+// RPC calls.
+//
+// This class is NOT part of the JavaScript class heirarchy (it has no JSG_RESOURCE_TYPE block),
+// it's only a C++ class used to abstract how to get a capnp client out of the object.
+class JsRpcClientProvider: public jsg::Object {
+public:
+  // Get a capnp client that can be used to dispatch one call.
+  //
+  // If this isn't the root object (i.e. this is a JsRpcProperty), the property path starting from
+  // the root object will be appended to `path`.
+  virtual rpc::JsRpcTarget::Client getClientForOneCall(
+      jsg::Lock& js, kj::Vector<kj::StringPtr>& path) = 0;
+};
+
+class JsRpcProperty;
+
+// Represents the promise returned by calling an RPC method. We don't use a regular Promise object,
+// but rather our own custom thenable, so that we can support pipelining on it.
+class JsRpcPromise: public JsRpcClientProvider {
+public:
+  JsRpcPromise(jsg::JsRef<jsg::JsPromise> inner,
+      IoOwn<rpc::JsRpcTarget::CallResults::Pipeline> pipeline)
+      : inner(kj::mv(inner)), pipeline(kj::mv(pipeline)) {}
+
+  rpc::JsRpcTarget::Client getClientForOneCall(
+      jsg::Lock& js, kj::Vector<kj::StringPtr>& path) override;
+
+  // Implement standard Promise interface, especially `then()` so that this works as a custom
+  // thenable.
+  //
+  // Note that we intentionally return jsg::JsValue rather than jsg::JsPromise because we actually
+  // do not want the JSG glue to recognize we're returning a promise triggering behavior that pins
+  // the JsRpcPromise in memory until it resolves. It's actually fine if the JsRpcPromise is GC'd
+  // before the promise resolves.
+  jsg::JsValue then(jsg::Lock& js, v8::Local<v8::Function> handler,
+      jsg::Optional<v8::Local<v8::Function>> errorHandler);
+  jsg::JsValue catch_(jsg::Lock& js, v8::Local<v8::Function> errorHandler);
+  jsg::JsValue finally(jsg::Lock& js, v8::Local<v8::Function> onFinally);
+
+  // Get a nested property, using pipelining.
+  kj::Maybe<jsg::Ref<JsRpcProperty>> getProperty(jsg::Lock& js, kj::String name);
+
+  JSG_RESOURCE_TYPE(JsRpcPromise) {
+    JSG_WILDCARD_PROPERTY(getProperty);
+    JSG_METHOD(then);
+    JSG_METHOD_NAMED(catch, catch_);
+    JSG_METHOD(finally);
+  }
+
```

Comment:
This should implement the `visitForMemoryInfo(...)` tracking the held promise

Reply:
Will do that at the same time as visitForGc.